### PR TITLE
implement logic for regular NodesRequests

### PR DESCRIPTION
**Pull request by zugz**
This makes a start on implementing the actual operation of the DHT, though
it's all wholly pure for now.

This has required some reworking of existing modules, and the addition of an
abstraction layer (NodeList).

I've also ended up adding to the dependencies, and using some GHC extensions.

Please review critically.

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/hs-toxcore/122)

### CODE REVIEW

**Comment by iphydf** [comment]
I'll review this in more detail later. I'm quite busy these two weeks. @robinlinden can you also review this?

---

Review status: 0 of 16 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[hstox.cabal, line 41 at r1](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-Kc9hXNdX5tXl0eB23Mn:-Kc9hXNdX5tXl0eB23Mo:boly6nr) ([raw file](https://github.com/toktok/hs-toxcore/blob/7b8b61c65906284e9302219fb521f61444a09968/hstox.cabal#L41)):*
> ```
>     , integer-gmp
>     , iproute
>     , MonadRandom               >= 0.5
> ```

What's in 0.5 that we need?

---

*[src/testsuite/Network/Tox/DHT/ClientListSpec.hs, line 50 at r1](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-Kc9hDGtYdhxPmCNskeP:-Kc9hDGtYdhxPmCNskeQ:bsj8cve) ([raw file](https://github.com/toktok/hs-toxcore/blob/7b8b61c65906284e9302219fb521f61444a09968/src/testsuite/Network/Tox/DHT/ClientListSpec.hs#L50)):*
> ```Haskell
> 
>   it "adding a non-viable node has no effect" $
>     property $ \(clientList::ClientList) time nodeInfo ->
> ```

Is this explicit type actually necessary?

---

*[src/tox/Network/Tox/Time.lhs, line 14 at r1](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-Kc9i-WSqLvBUJ3AqFPU:-Kc9i-WSqLvBUJ3AqFPV:bjg5mfc) ([raw file](https://github.com/toktok/hs-toxcore/blob/7b8b61c65906284e9302219fb521f61444a09968/src/tox/Network/Tox/Time.lhs#L14)):*
> ```lhaskell
> 
> -- | TODO
> type TimeStamp = Integer
> ```

Do timestamps need to be arbitrary size integers? Would Int64 be enough? Also, "timestamp" is one word, not "time stamp".

---

*[src/tox/Network/Tox/DHT/DhtState.lhs, line 163 at r1](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-Kc9imvjyiUw404V65lF:-Kc9imvjyiUw404V65lG:b-4frsg2) ([raw file](https://github.com/toktok/hs-toxcore/blob/7b8b61c65906284e9302219fb521f61444a09968/src/tox/Network/Tox/DHT/DhtState.lhs#L163)):*
> ```lhaskell
> traverseNodeLists :: Applicative f =>
>   (forall l. NodeList l => l -> f l) -> DhtState -> f DhtState
> traverseNodeLists f dhtState@DhtState{ dhtCloseList, dhtSearchList }  =
> ```

nit: only one space before `=`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-:-Kc9mu33aPtQJYXJFIg9:b8zdmvl)*

File: src/tox/Network/Tox/Time.lhs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+
+{-------------------------------------------------------------------------------
+ -
+ - :: Tests.
+ -
+ ------------------------------------------------------------------------------}
+
+instance Arbitrary Timestamp
+  where arbitrary = (Timestamp <$>) $ Clock.TimeSpec <$> arbitrary <*> arbitrary
```
Comment (pippijn) [inline src/tox/Network/Tox/Time.lhs]: `<$>` and `<*>` need explicit import for GHC <7.10.

**Comment by iphydf** [comment]
:lgtm_strong:

---

Reviewed 6 of 16 files at r1, 10 of 10 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-:-Kd5W-KyaMKEsl24skAk:b-jbsf8z)*

**Comment by zugz** [comment]
* Saturday, 2017-02-04 at 12:23 -0800 - iphydf <notifications@github.com>:

> ---
> 
> *[hstox.cabal, line 41 at r1](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-Kc9hXNdX5tXl0eB23Mn:-Kc9hXNdX5tXl0eB23Mo:boly6nr) ([raw file](https://github.com/toktok/hs-toxcore/blob/7b8b61c65906284e9302219fb521f61444a09968/hstox.cabal#L41)):*
> > ```
> >     , integer-gmp
> >     , iproute
> >     , MonadRandom               >= 0.5
> > ```
> 
> What's in 0.5 that we need?
> 
> ---

'uniform', for picking a random element of a list. But probably it makes
more sense to add a simple implementation and not have the version
restriction; I'll do that.

> *[src/testsuite/Network/Tox/DHT/ClientListSpec.hs, line 50 at r1](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-Kc9hDGtYdhxPmCNskeP:-Kc9hDGtYdhxPmCNskeQ:bsj8cve) ([raw file](https://github.com/toktok/hs-toxcore/blob/7b8b61c65906284e9302219fb521f61444a09968/src/testsuite/Network/Tox/DHT/ClientListSpec.hs#L50)):*
> > ```Haskell
> > 
> >   it "adding a non-viable node has no effect" $
> >     property $ \(clientList::ClientList) time nodeInfo ->
> > ```
> 
> Is this explicit type actually necessary?

It was in an earlier version, but not any more. Done.

> *[src/tox/Network/Tox/Time.lhs, line 14 at r1](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-Kc9i-WSqLvBUJ3AqFPU:-Kc9i-WSqLvBUJ3AqFPV:bjg5mfc) ([raw file](https://github.com/toktok/hs-toxcore/blob/7b8b61c65906284e9302219fb521f61444a09968/src/tox/Network/Tox/Time.lhs#L14)):*
> > ```lhaskell
> > 
> > -- | TODO
> > type TimeStamp = Integer
> > ```
> 
> Do timestamps need to be arbitrary size integers? Would Int64 be enough? Also, "timestamp" is one word, not "time stamp".

Setting it to Integer was just temporary while I decided what time
package to use; it'll be System.Clock.TimeSpec in the future.

Done s/TimeStamp/Timestamp/g.

> *[src/tox/Network/Tox/DHT/DhtState.lhs, line 163 at r1](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-Kc9imvjyiUw404V65lF:-Kc9imvjyiUw404V65lG:b-4frsg2) ([raw file](https://github.com/toktok/hs-toxcore/blob/7b8b61c65906284e9302219fb521f61444a09968/src/tox/Network/Tox/DHT/DhtState.lhs#L163)):*
> > ```lhaskell
> > traverseNodeLists :: Applicative f =>
> >   (forall l. NodeList l => l -> f l) -> DhtState -> f DhtState
> > traverseNodeLists f dhtState@DhtState{ dhtCloseList, dhtSearchList }  =
> > ```
> 
> nit: only one space before `=`.

Done.

**Comment by zugz** [comment]
* Tuesday, 2017-02-14 at 06:20 -0800 - Pippijn van Steenhoven <notifications@github.com>:

> pippijn commented on this pull request.
> 
> `<$>` and `<*>` need explicit import for GHC <7.10.

Yes, thanks, there'll be a few of these. Will fix. I do wish ghc-7.10
had an option to warn about such incompatibilities...

**Comment by iphydf** [comment]
:lgtm_strong: generally happy with this, just a small comment on Safe Haskell.

---

Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[src/tox/Network/Tox/DHT/Operation.lhs, line 7 at r3](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-KduPfJ_hagwvMZz4x08:-KduPfJ_hagwvMZz4x09:b-lz1uvo) ([raw file](https://github.com/toktok/hs-toxcore/blob/9d8187cbed1c5f731c8ccc6801e11c8e95cff049/src/tox/Network/Tox/DHT/Operation.lhs#L7)):*
> ```lhaskell
> {-# LANGUAGE MultiParamTypeClasses #-}
> {-# LANGUAGE NamedFieldPuns        #-}
> {-# LANGUAGE Safe                  #-}
> ```

Control.Monad.Random is not Safe Haskell in earlier versions. Try to use Trustworthy here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/hs-toxcore/122#-:-KduOCMUumjCQwH_c4La:bs79mgb)*

**Pull request merged**